src/common/hist_util.cc

```diff
@@ -785,41 +860,46 @@ void BuildHistSparseKernel(const common::Span<const size_t> rid_span, const floa
       const size_t icol_end_prefect = row_ptr[rid[i+Prefetch::kPrefetchOffset]+1];
 
       PREFETCH_READ_T0(pgh + 2*rid[i + Prefetch::kPrefetchOffset]);
-      for (size_t j = icol_start_prftch; j < icol_end_prefect; j+=Prefetch::kPrefetchStep) {
+      for (size_t j = icol_start_prftch; j < icol_end_prefect; j+=prefetch_step) {
         PREFETCH_READ_T0(gradient_index + j);
       }
     }
-
+    size_t jd = 0;
     for (size_t j = icol_start; j < icol_end; ++j) {
-      const uint32_t idx_bin = 2 * gradient_index[j];
+      const uint32_t idx_bin = 2 * ((uint32_t)(gradient_index[j]) + disps[jd]);
+      ++jd;
       hist_data[idx_bin]   += pgh[idx_gh];
       hist_data[idx_bin+1] += pgh[idx_gh+1];
     }
   }
 }
 
-template<typename FPType, bool do_prefetch>
+template<typename FPType, bool do_prefetch, typename T>
 void BuildHistKernel(const common::Span<const size_t> rid_span, const float* pgh,
-                     const uint32_t* gradient_index,  const size_t* row_ptr, const bool isDense,
+                     const T* gradient_index, const uint32_t* disps, const size_t* row_ptr, const bool isDense,
                      FPType* hist_data) {
   if (rid_span.size() && isDense) {
     const size_t n_features = row_ptr[rid_span[0]+1] - row_ptr[rid_span[0]];
-    BuildHistDenseKernel<FPType, do_prefetch>(rid_span, pgh, gradient_index,
+    BuildHistDenseKernel<FPType, do_prefetch>(rid_span, pgh, gradient_index, disps,
                                               n_features, hist_data);
   } else {
-    BuildHistSparseKernel<FPType, do_prefetch>(rid_span, pgh, gradient_index, row_ptr, hist_data);
+    BuildHistSparseKernel<FPType, do_prefetch>(rid_span, pgh, gradient_index, disps, row_ptr, hist_data);
   }
 }
 
+template<typename T>
 void GHistBuilder::BuildHist(const std::vector<GradientPair>& gpair,
                              const RowSetCollection::Elem row_indices,
-                             const GHistIndexMatrix& gmat,
+                             //const GHistIndexMatrix& gmat,
+                             const T* index,
+                             const uint32_t* disps,
```

Comment:
Please, use common::Span instead of row pointers where it's possible.

Reply:
done